### PR TITLE
Ensure Theater audio and visual files start playback at the same time.

### PR DESCRIPTION
**Pull request by jmkulwik**
Added wait for both the audio and the video file to be loaded before starting playback in the Theater.

Prior to this change, each would play as it loaded. The synchronization still isn't perfect, but is much closer.
Future work: To get perfect synchronization, we'd likely need to change from separate audio/visual generation (gif and wav) to a combined file such as an mp4.

By adjusting the `visibility` css field, we also fix the issue where a broken image symbol showed up whenever the run button was clicked in theater levels.

Before (with sound): 

https://user-images.githubusercontent.com/8324574/131734330-d439b0e0-aa7c-49fb-b45e-e49b1ec535ab.mp4

After (with sound): 

https://user-images.githubusercontent.com/8324574/131734351-6c8499f4-8e69-45f0-8db0-784f1ed5069e.mp4

## Links

- Broken image symbol: [CSA-599](https://codedotorg.atlassian.net/browse/CSA-599)
- A/V Playback: [CSA-792](https://codedotorg.atlassian.net/browse/CSA-792)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by molly-moen — APPROVED** [review]

**Review by sanchitmalhotra126 — APPROVED** [review]
Nice! This may be a follow-up, but do we also need to account for load time once we receive both URLs and make sure they're in sync? (there are `canplay`/`canplaythrough` events we could use)

**Comment by jmkulwik** [comment]
Added in a forced wait for both files to load

File: apps/src/javalab/Theater.js [hunk 1]
```diff
@@ -7,24 +7,45 @@ export default class Theater {
...
+    this.visualResponse = null;
   }
 
   handleSignal(data) {
     switch (data.value) {
       case TheaterSignalType.AUDIO_URL: {
-        this.getAudioElement().src = data.detail.url;
+        // Wait for the audio to load before starting playback
+        this.audioResponse = data.detail.url;
+        this.getAudioElement().src = this.audioResponse;
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/Theater.js]: nit: do we need to save off the url to  `this.audioResponse`?

File: apps/src/javalab/Theater.js [hunk 2]
```diff
@@ -7,24 +7,45 @@ export default class Theater {
...
+        this.getAudioElement().oncanplaythrough = () => this.startPlayback();
         break;
       }
       case TheaterSignalType.VISUAL_URL: {
-        this.getImgElement().src = data.detail.url;
+        // Preload the image. Once it's ready, start the playback
+        this.visualResponse = data.detail.url;
+        var img = new Image();
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/Theater.js]: nit: `const`?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

As a future follow-up, I was also thinking we can have Javabuilder just send a single response for every theater update that has both audio and visual URLs as part of the details object. Since we always send both URLs together, it would help simplify the logic here a bit.

File: apps/src/javalab/Theater.js [hunk 3]
```diff
@@ -7,25 +7,41 @@ export default class Theater {
...
         break;
     }
   }
 
+  startPlayback() {
+    this.loadEventsFinished++;
+    // We expect exactly 2 responses from Javabuilder. One for audio and one for video.
+    // Wait for both to respond and load before starting playback.
+    if (this.loadEventsFinished > 1) {
+      this.getImgElement().style.visibility = 'visible';
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/Theater.js:37]: Curious, does toggling the visibility fix the broken image issue too?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

File: apps/src/javalab/Theater.js (hunk 3 above)
Comment (jmkulwik) [reply apps/src/javalab/Theater.js:37]: Yes, it does! Good eye. Thanks for the reminder too - I forgot to include that in the description :)

**Comment by jmkulwik** [comment]
Successful drone run: https://drone.cdn-code.org/code-dot-org/code-dot-org/10010

**Pull request merged**